core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java

```diff
@@ -0,0 +1,1001 @@
+
+    @VisibleForTesting
+    @Nullable
+    Class<?> containerType() {
+        // 'List' or 'Set'
+        return containerType;
+    }
+
+    @VisibleForTesting
+    Class<?> elementType() {
+        return elementType;
+    }
+
+    @VisibleForTesting
+    @Nullable
+    Object defaultValue() {
+        return defaultValue;
+    }
+
+    AggregationStrategy aggregationStrategy() {
+        return aggregationStrategy;
+    }
+
+    boolean hasContainer() {
+        return containerType != null &&
+               (List.class.isAssignableFrom(containerType) || Set.class.isAssignableFrom(containerType));
+    }
+
+    Object resolve(ResolverContext ctx) {
+        final Object resolved = resolver.apply(this, ctx);
+        return shouldWrapValueAsOptional ? Optional.ofNullable(resolved)
+                                         : resolved;
+    }
+
+    private static Object convert(String value, Class<?> elementType,
+                                  @Nullable EnumConverter<?> enumConverter) {
+        return enumConverter != null ? enumConverter.toEnum(value)
+                                     : stringToType(value, elementType);
+    }
+
+    @Nullable
+    private Object convert(@Nullable String value) {
+        if (value == null) {
+            return defaultOrException();
+        }
+        return convert(value, elementType, enumConverter);
+    }
+
+    @Nullable
+    private Object defaultOrException() {
+        if (!shouldExist) {
+            // May return 'null' if no default value is specified.
+            return defaultValue;
+        }
+        throw new IllegalArgumentException("Mandatory parameter is missing: " + httpElementName);
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                          .add("annotation",
+                               annotation != null ? annotation.annotationType().getSimpleName() : "(none)")
+                          .add("httpElementName", httpElementName)
+                          .add("pathVariable", isPathVariable)
+                          .add("shouldExist", shouldExist)
+                          .add("shouldWrapValueAsOptional", shouldWrapValueAsOptional)
+                          .add("elementType", elementType.getSimpleName())
+                          .add("containerType",
+                               containerType != null ? containerType.getSimpleName() : "(none)")
+                          .add("defaultValue", defaultValue)
+                          .add("defaultValueType",
+                               defaultValue != null ? defaultValue.getClass().getSimpleName() : "(none)")
+                          .add("resolver", resolver)
+                          .add("enumConverter", enumConverter)
+                          .toString();
+    }
+
+    static Builder builder(AnnotatedElement annotatedElement, Type type) {
+        return new Builder(annotatedElement, type);
+    }
+
+    static class Builder {
```

Comment:
I added a builder for `AnnotatedValueResolver` to make each annotation's specification clear like https://github.com/line/armeria/pull/1191/files#diff-9df84bbcae0217f0942af4e4d6a3f8c0R244.